maskrcnn_benchmark/structures/segmentation_mask.py

```diff
@@ -8,68 +11,213 @@
-            dim = height
-            idx = 1
+            reshaped_contour = []
+            for entity in contour:
+                assert len(entity.shape) == 3
+                assert entity.shape[1] == 1, \
+                    'Hierarchical contours are not allowed'
+                reshaped_contour.append(entity.reshape(-1).tolist())
+            contours.append(reshaped_contour)
+        return contours
 
-        flip_idx = list(range(dim)[::-1])
-        flipped_masks = self.masks.index_select(dim, flip_idx)
-        return Mask(flipped_masks, self.size, self.mode)
 
-    def crop(self, box):
-        w, h = box[2] - box[0], box[3] - box[1]
+    def __len__(self):
+        return len(self.masks)
 
-        cropped_masks = self.masks[:, box[1] : box[3], box[0] : box[2]]
-        return Mask(cropped_masks, size=(w, h), mode=self.mode)
 
-    def resize(self, size, *args, **kwargs):
-        pass
+    def __getitem__(self, index):
+        # Probably it can cause some overhead
+        # but preserves consistency
+        masks = self.masks[index].clone()
+        return BinaryMaskList(masks, self.size)
+
 
+    def __iter__(self):
+        return iter(self.masks)
 
-class Polygons(object):
-    """
+
+    def __repr__(self):
+        s = self.__class__.__name__ + "("
+        s += "num_instances={}, ".format(len(self.masks))
+        s += "image_width={}, ".format(self.size[0])
+        s += "image_height={})".format(self.size[1])
+        return s
+
+
+
+class PolygonInstance(object):
+    '''
     This class holds a set of polygons that represents a single instance
     of an object mask. The object can be represented as a set of
     polygons
-    """
+    '''
+
+    def __init__(self, polygons, size):
+        '''
+            Arguments:
+                a list of lists of numbers.
+                The first level refers to all the polygons that compose the
+                object, and the second level to the polygon coordinates.
+        '''
+        if isinstance(polygons, (list, tuple)):
+            valid_polygons = []
+            for p in polygons:
+                p = torch.as_tensor(p, dtype=torch.float32)
+                if len(p) >= 6: # 3 * 2 coordinates
+                    valid_polygons.append(p)
+            polygons = valid_polygons
+
+        elif isinstance(polygons, PolygonInstance):
+            polygons = polygons.polygons.copy()
 
-    def __init__(self, polygons, size, mode):
-        # assert isinstance(polygons, list), '{}'.format(polygons)
-        if isinstance(polygons, list):
-            polygons = [torch.as_tensor(p, dtype=torch.float32) for p in polygons]
-        elif isinstance(polygons, Polygons):
-            polygons = polygons.polygons
+        else:
+            RuntimeError('Type of argument `polygons` is not allowed:%s'\
+                %(type(polygons)))
+
+        ''' This crashes the training way too many times...
+        for p in polygons:
+            assert p[::2].min() >= 0
```

Comment:
can you explain why the training crashes with those asserts? are the elements out of bounds or negatives?

Reply:
both cases happen in the COCO training.
Ideally, this assert should not be a problem, however it seems that when the polygons are converted to binary masks, the implementation handles them without throwing warnings.

Reply:
I still don't understand this correctly. This assumes that `p` is a polygon, right?

Reply:
@fmassa please take a look at the `PolygonInstance.crop` method, which is a bit odd to me: instead of clamping (as noted in the in-line comment) it just subtracts the left/top boundrary of the cropping box of the polygons, and passes the reduced cropping size as an argument for the new `PolygonInstance` instance.

This was in the original implementation, and I suppose it heavily relies on the fact that `polygon->binary` conversion will handle this issue by just ignoring over- and under-flowing polygon coordinates. So this could be why it is working without, and crashing with these asserts

Reply:
Oh I see, yes, it makes sense.

The cropping was implemented that way to mimic the implementation in Detectron for the fused `project_on_box`. Actually, to make things really correct we would need to identify that a point is negative, and add new edges to the polygon. Just clamping the coordinates of the polygon is unfortunately not enough